File: app/recurrentes.py
"""Gastos recurrentes: qué se está pagando todos los meses sin mirarlo.

Espejo en Python de web/js/recurrentes.js. Se duplica por el mismo motivo que
el termómetro: la web lee Supabase directo y el bot no puede depender de que
la web esté abierta, ni al revés.

QUÉ CUENTA COMO RECURRENTE — tres condiciones, y las tres hacen falta:

  1. Al menos 3 cargos del mismo ítem. Con dos no hay periodicidad, hay una
     coincidencia.
  2. Espaciados regulares: la mediana de los días entre cargos tiene que
     parecerse a un período conocido, y ningún intervalo puede despegarse
     mucho de ella.
  3. Monto estable respecto de la mediana.

La tercera es la que separa un gasto RECURRENTE de uno FRECUENTE. El súper
también aparece todos los meses, pero por montos distintos: eso es un hábito,
no un débito automático.

LO QUE EL BOT HACE Y LO QUE NO
Sugiere revisar. No cancela nada, no puede cancelar nada, y no dice que haya
que cancelar: no sabe si el gimnasio al que no vas es un olvido o una decisión.
Muestra hace cuánto se paga y cuánto se lleva acumulado, que es el dato que
nadie tiene a mano, y ahí se detiene.
"""

from __future__ import annotations

import logging
from datetime import date
from decimal import Decimal, InvalidOperation

logger = logging.getLogger(__name__)

CARGOS_MINIMOS = 3
TOLERANCIA_DIAS = Decimal("0.35")
TOLERANCIA_MONTO = Decimal("0.25")

# (nombre, días, cuántas veces por mes)
PERIODOS = (
    ("semanal", Decimal("7"), Decimal("4.33")),
    ("quincenal", Decimal("15"), Decimal("2")),
    ("mensual", Decimal("30.44"), Decimal("1")),
    ("bimestral", Decimal("61"), Decimal("0.5")),
    ("trimestral", Decimal("91"), Decimal("0.33")),
    ("anual", Decimal("365"), Decimal("1") / 12),
)


class Recurrente:
    """Un cargo que se repite con monto y período estables."""

    __slots__ = (
        "clave", "categoria", "cargos", "periodo", "monto_tipico", "por_mes",
        "primero", "ultimo", "dias_desde_ultimo", "total_pagado", "meses_activo",
    )

    def __init__(self, **kw) -> None:
        for campo in self.__slots__:
            setattr(self, campo, kw.get(campo))


def _mediana(valores: list[Decimal]) -> Decimal:
    if not valores:
        return Decimal("0")
    ordenados = sorted(valores)
    medio = len(ordenados) // 2
    if len(ordenados) % 2:
        return ordenados[medio]
    return (ordenados[medio - 1] + ordenados[medio]) / 2


def _periodo_de(dias: Decimal) -> tuple[str, Decimal, Decimal] | None:
    """El período conocido más parecido, o None si no se parece a ninguno."""
    mejor, distancia = None, Decimal("999")
    for nombre, largo, por_mes in PERIODOS:
        d = abs(dias - largo) / largo
        if d < distancia:
            mejor, distancia = (nombre, largo, por_mes), d
    return mejor if distancia <= TOLERANCIA_DIAS else None


def _clave_de(fila: dict) -> str:
    """La del termómetro si está; si no, lo que haya."""
    for campo in ("clave_item", "comercio", "descripcion", "categoria"):
        valor = (fila.get(campo) or "").strip().lower()
        if valor:
            return valor
    return ""


def detectar(filas: list[dict], moneda: str, hoy: date | None = None) -> list[Recurrente]:
    """Los cargos recurrentes de una moneda, del que más pesa al que menos."""
    hoy = hoy or date.today()

    por_item: dict[str, list[tuple[date, Decimal, str]]] = {}
    for fila in filas:
        if fila.get("tipo") != "gasto" or fila.get("moneda") != moneda:
            continue
        clave = _clave_de(fila)
        if not clave:
            continue
        try:
            cuando = date.fromisoformat(fila["fecha"])
            monto = Decimal(str(fila["monto"]))
        except (KeyError, ValueError, TypeError, InvalidOperation):
            continue
        if monto <= 0:
            continue
        por_item.setdefault(clave, []).append((cuando, monto, fila.get("categoria") or ""))

    salida: list[Recurrente] = []

    for clave, cargos in por_item.items():
        if len(cargos) < CARGOS_MINIMOS:
            continue
        cargos.sort(key=lambda c: c[0])

        intervalos = [
            Decimal((cargos[i][0] - cargos[i - 1][0]).days)
            for i in range(1, len(cargos))
            if (cargos[i][0] - cargos[i - 1][0]).days > 0
        ]
        if len(intervalos) < CARGOS_MINIMOS - 1:
            continue

        dias_mediana = _mediana(intervalos)
        if dias_mediana <= 0:
            continue
        if any(abs(d - dias_mediana) / dias_mediana > TOLERANCIA_DIAS for d in intervalos):
            continue

        periodo = _periodo_de(dias_mediana)
        if periodo is None:
            continue
        nombre, largo, por_mes = periodo

        montos = [c[1] for c in cargos]
        tipico = _mediana(montos)
        if tipico <= 0:
            continue
        if any(abs(m - tipico) / tipico > TOLERANCIA_MONTO for m in montos):
            continue

        dias_desde = (hoy - cargos[-1][0]).days

        salida.append(
            Recurrente(
                clave=clave,
                categoria=cargos[-1][2],
                cargos=len(cargos),
                periodo=nombre,
                monto_tipico=tipico.quantize(Decimal("0.01")),
                por_mes=(tipico * por_mes).quantize(Decimal("0.01")),
                primero=cargos[0][0],
                ultimo=cargos[-1][0],
                dias_desde_ultimo=dias_desde,
                total_pagado=sum(montos, Decimal("0")),
                meses_activo=round((cargos[-1][0] - cargos[0][0]).days / 30.44),
            )
        )

    salida.sort(key=lambda r: r.por_mes, reverse=True)
    return salida


def total_mensual(recurrentes: list[Recurrente]) -> Decimal:
    return sum((r.por_mes for r in recurrentes), Decimal("0"))


def redactar(recurrentes: list[Recurrente], moneda, formatear_monto) -> str:
    """El mensaje del bot. Informa y sugiere revisar; no aconseja cancelar."""
    if not recurrentes:
        return (
            "Todavía no detecté gastos recurrentes 🔍\n\n"
            "Necesito al menos 3 cargos del mismo ítem, espaciados parejo y por "
            "montos parecidos. Con más meses cargados esto mejora bastante."
        )

    total = total_mensual(recurrentes)
    lineas = [
        f"🔁 Tenés {len(recurrentes)} "
        f"{'gasto recurrente' if len(recurrentes) == 1 else 'gastos recurrentes'}, "
        f"{formatear_monto(total, moneda)} por mes en total:",
        "",
    ]

    for r in recurrentes[:8]:
        lineas.append(
            f"• {r.clave}: {formatear_monto(r.monto_tipico, moneda)} {r.periodo} "
            f"({r.cargos} cargos)"
        )

    if len(recurrentes) > 8:
        lineas.append(f"…y {len(recurrentes) - 8} más.")

    # La sugerencia: los que llevan mucho tiempo cobrándose. Se los nombra y se
    # invita a mirarlos, sin decir qué hacer.
    veteranos = [r for r in recurrentes if r.meses_activo >= 6]
    if veteranos:
        lineas.append("")
        lineas.append("Estos vienen de hace rato, por si querés revisarlos:")
        for r in veteranos[:3]:
            lineas.append(
                f"  {r.clave} — {r.meses_activo} meses, "
                f"{formatear_monto(r.total_pagado, moneda)} acumulados"
            )
        lineas.append("Los dejo a la vista nomás: si los usás, están bien donde están.")

    lineas.append("")
    lineas.append(
        "ℹ️ Esto mejora con el tiempo: cuantos más meses tengas cargados, "
        "mejor distingo una suscripción de una coincidencia."
    )
    return "\n".join(lineas)

File: web/js/recurrentes.js
// Detección de gastos recurrentes y suscripciones.

const CARGOS_MINIMOS = 3;
const TOLERANCIA_DIAS = 0.35;
const TOLERANCIA_MONTO = 0.25;
const PERIODOS = [
  { nombre: "semanal", dias: 7, porMes: 4.33 },
  { nombre: "quincenal", dias: 15, porMes: 2 },
  { nombre: "mensual", dias: 30.44, porMes: 1 },
  { nombre: "bimestral", dias: 61, porMes: 0.5 },
  { nombre: "trimestral", dias: 91, porMes: 0.33 },
  { nombre: "anual", dias: 365, porMes: 1 / 12 },
];

function mediana(valores) {
  if (!valores.length) return 0;
  const o = [...valores].sort((a, b) => a - b);
  const m = Math.floor(o.length / 2);
  return o.length % 2 ? o[m] : (o[m - 1] + o[m]) / 2;
}

function periodoDe(diasMediana) {
  let mejor = null;
  let distancia = Infinity;
  for (const p of PERIODOS) {
    const d = Math.abs(diasMediana - p.dias) / p.dias;
    if (d < distancia) [mejor, distancia] = [p, d];
  }
  return distancia <= TOLERANCIA_DIAS ? mejor : null;
}

/** Clave de agrupación: la del termómetro si está, si no la descripción. */
function claveDe(m) {
  return (m.clave_item || m.comercio || m.descripcion || m.categoria || "").trim().toLowerCase();
}

export function detectar(movimientos, moneda, hoyISO) {
  const porItem = new Map();

  for (const m of movimientos ?? []) {
    if (m.tipo !== "gasto" || m.moneda !== moneda) continue;
    const clave = claveDe(m);
    if (!clave) continue;
    const monto = Number(m.monto);
    if (!Number.isFinite(monto) || monto <= 0) continue;
    if (!porItem.has(clave)) porItem.set(clave, []);
    porItem.get(clave).push({ fecha: m.fecha, monto, categoria: m.categoria ?? "" });
  }

  const hoy = new Date(hoyISO ?? new Date().toISOString().slice(0, 10));
  const salida = [];

  for (const [clave, cargos] of porItem) {
    if (cargos.length < CARGOS_MINIMOS) continue;
    cargos.sort((a, b) => a.fecha.localeCompare(b.fecha));

    const intervalos = [];
    for (let i = 1; i < cargos.length; i++) {
      const dias = (new Date(cargos[i].fecha) - new Date(cargos[i - 1].fecha)) / 86400000;
      if (dias > 0) intervalos.push(dias);
    }
    if (intervalos.length < CARGOS_MINIMOS - 1) continue;

    const diasMediana = mediana(intervalos);
    if (diasMediana <= 0) continue;

    const regular = intervalos.every(
      (d) => Math.abs(d - diasMediana) / diasMediana <= TOLERANCIA_DIAS
    );
    if (!regular) continue;

    const periodo = periodoDe(diasMediana);
    if (!periodo) continue;

    const montos = cargos.map((c) => c.monto);
    const montoTipico = mediana(montos);
    if (montoTipico <= 0) continue;

    const estable = montos.every(
      (x) => Math.abs(x - montoTipico) / montoTipico <= TOLERANCIA_MONTO
    );
    if (!estable) continue;

    const ultimo = cargos[cargos.length - 1];
    const diasDesdeUltimo = Math.round((hoy - new Date(ultimo.fecha)) / 86400000);

    salida.push({
      clave,
      categoria: ultimo.categoria,
      cargos: cargos.length,
      periodo: periodo.nombre,
      montoTipico,
      porMes: montoTipico * periodo.porMes,
      primero: cargos[0].fecha,
      ultimo: ultimo.fecha,
      diasDesdeUltimo,
      totalPagado: montos.reduce((t, x) => t + x, 0),
      vencido: diasDesdeUltimo > periodo.dias * (1 + TOLERANCIA_DIAS * 2),
      mesesActivo: Math.round(
        (new Date(ultimo.fecha) - new Date(cargos[0].fecha)) / 86400000 / 30.44
      ),
    });
  }

  return salida.sort((a, b) => b.porMes - a.porMes);
}

/** Lo que representan todos juntos por mes. */
export function totalMensual(recurrentes) {
  return (recurrentes ?? []).reduce((t, r) => t + r.porMes, 0);
}
